arangod/Aql/IResearchViewExecutor.tpp

```diff
@@ -1412,122 +1439,265 @@ bool IResearchViewHeapSortExecutor<ExecutionTraits>::fillBufferInternal(
-        break;
-      }
-      continue;
+      return gotData;
     }
     if constexpr (Base::isMaterialized) {
       // The collection must stay the same for all documents in the buffer
-      TRI_ASSERT(_segment == &this->_reader->segment(_readerOffset));
+      TRI_ASSERT(reader.segment ==
+                 &this->_reader->segment(reader.readerOffset));
       if (!documentId.isSet()) {
         // No document read, we cannot write it.
         continue;
       }
     }
+    auto current = bufferIdx.load(std::memory_order_relaxed);
+    if constexpr (parallel) {
+      while (!bufferIdx.compare_exchange_weak(current, current + 1)) {
+      }
+    }
+    auto& viewSegment = this->_reader->segment(reader.readerOffset);
     if constexpr (Base::isLateMaterialized) {
-      this->_indexReadBuffer.pushValue(this->_reader->segment(_readerOffset),
-                                       _doc->value);
+      if constexpr (parallel) {
+        this->_indexReadBuffer.setValue(current, viewSegment,
+                                        reader.doc->value);
+      } else {
+        this->_indexReadBuffer.pushValue(viewSegment, reader.doc->value);
+      }
     } else {
-      this->_indexReadBuffer.pushValue(documentId);
+      if constexpr (parallel) {
+        this->_indexReadBuffer.setValue(current, viewSegment, documentId);
+      } else {
+        this->_indexReadBuffer.pushValue(viewSegment, documentId);
+      }
     }
+    --reader.atMost;
     gotData = true;
 
     if constexpr (ExecutionTraits::EmitSearchDoc) {
       TRI_ASSERT(this->infos().searchDocIdRegId().isValid());
-      this->_indexReadBuffer.pushSearchDoc(
-          _doc->value, this->_reader->segment(_readerOffset));
+      if constexpr (parallel) {
+        this->_indexReadBuffer.setSearchDoc(current, viewSegment,
+                                            reader.doc->value);
+      } else {
+        this->_indexReadBuffer.pushSearchDoc(reader.doc->value, viewSegment);
+      }
     }
 
     // in the ordered case we have to write scores as well as a document
     if constexpr (ExecutionTraits::Ordered) {
-      // Writes into _scoreBuffer
-      this->fillScores(*_scr);
+      if constexpr (parallel) {
+        (*reader.scr)(this->_indexReadBuffer.getScoreBuffer(
+            current * this->infos().scoreRegistersCount()));
+      } else {
+        this->fillScores(*reader.scr);
+      }
     }
 
     if constexpr (Base::usesStoredValues) {
-      TRI_ASSERT(_doc);
-      this->pushStoredValues(*_doc);
+      TRI_ASSERT(reader.doc);
+      TRI_ASSERT(std::distance(_segmentReaders.data(), &reader) <
+                 static_cast<ptrdiff_t>(_segmentReaders.size()));
+      this->template pushStoredValues<parallel>(
+          *reader.doc, std::distance(_segmentReaders.data(), &reader), current);
+    }
+    if constexpr (!parallel) {
+      // doc and scores are both pushed, sizes must now be coherent
+      this->_indexReadBuffer.assertSizeCoherence();
```

Comment:
just pass parallel bool to assert function?

Reply:
The main point is avoid different in compile time code when possible

Reply:
Not sure if I got this. Could you elaborate a bit more? IMO just don't having call at all is always better than having `if(!parallel) return` inside.

Reply:
It's empty call, only asserts ibside

Reply:
Still let's leave it as is for now.